src/psyclone/psyir/transformations/omp_task_trans.py

```diff
@@ -0,0 +1,192 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2021, Science and Technology Facilities Council.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author A. B. G. Chalk STFC Daresbury Lab
+''' This module provides the OMPTaskTrans transformation.'''
+from psyclone.errors import GenerationError
+from psyclone.domain.common.transformations import KernelModuleInlineTrans
+from psyclone.psyGen import Kern
+from psyclone.psyir.transformations.fold_conditional_return_expressions_trans \
+        import FoldConditionalReturnExpressionsTrans
+from psyclone.psyir.transformations.parallel_loop_trans import\
+    ParallelLoopTrans
+from psyclone.psyir.nodes import CodeBlock, Call
+from psyclone.psyir.nodes import DynamicOMPTaskDirective
+from psyclone.psyir.transformations.inline_trans import InlineTrans
+from psyclone.psyir.transformations.transformation_error import \
+        TransformationError
+
+
+class OMPTaskTrans(ParallelLoopTrans):
```

Comment:
Please can you add this transformation to the user guide (they're in alpha-order in `transformations.rst`). If appropriate, you could document the limitation to `_update_node` in this class docstring (in which case it will appear in the User Guide) or alternatively, if it's too low level, in the section on Tasking in the Developer Guide.

Reply:
It makes me think that some of the design decisions I made here were a mistake, and that most of the logic for dependency computation should have been left in `OMPTaskDirective`, and inherited into `DynamicOMPTaskDirective` - this would allow the `OMPTaskDirective` to have an `_update_node` to call `_compute_clauses` again, wheras the `DynamicOMPTaskDirective` could just have an empty `_update_node` (to avoid computing clauses early).

I'm not sure I want to go through changing that for this PR unless you think it would be a significant improvement though at this point.

Reply:
No, let's capture it as an Issue and a TODO and leave it at that for now.

Reply:
Added